--- net/minecraft/src/EntityRenderer.java
+++ net/minecraft/src/EntityRenderer.java
@@ -10,6 +10,7 @@
 import org.lwjgl.opengl.GLContext;
 import org.lwjgl.opengl.NVFogDistance;
 import org.lwjgl.util.glu.GLU;
+import org.lwjgl.input.Keyboard;
 
 public class EntityRenderer {
 	private Minecraft mc;
@@ -261,6 +262,7 @@
 			this.mc.thePlayer.setAngles(var4, var5 * (float)var6);
 		}
 
+		LabyrinthModNative.Update();
 		if(!this.mc.skipRenderWorld) {
 			ScaledResolution var7 = new ScaledResolution(this.mc.displayWidth, this.mc.displayHeight);
 			int var8 = var7.getScaledWidth();
@@ -268,8 +270,14 @@
 			int var10 = Mouse.getX() * var8 / this.mc.displayWidth;
 			int var11 = var9 - Mouse.getY() * var9 / this.mc.displayHeight - 1;
 			if(this.mc.theWorld != null) {
+				if(!LabyrinthModNative.Loaded) {
+					LabyrinthModNative.Setup(this.mc);
+				}
 				this.renderWorld(var1);
 				this.mc.ingameGUI.renderGameOverlay(var1, this.mc.currentScreen != null, var10, var11);
+				if(!Keyboard.isKeyDown(Keyboard.KEY_F3)) {
+					LabyrinthModNative.Draw();
+				}
 			} else {
 				GL11.glViewport(0, 0, this.mc.displayWidth, this.mc.displayHeight);
 				GL11.glClearColor(0.0F, 0.0F, 0.0F, 0.0F);
@@ -279,6 +287,7 @@
 				GL11.glMatrixMode(GL11.GL_MODELVIEW);
 				GL11.glLoadIdentity();
 				this.setupOverlayRendering();
+				LabyrinthModNative.Reset();
 			}
 
 			if(this.mc.currentScreen != null) {
--- /dev/null
+++ net/minecraft/src/LabyrinthModNative.java
@@ -1,0 +1,94 @@
+package net.minecraft.src;
+
+import java.awt.DisplayMode;
+import java.lang.reflect.Field;
+import java.util.Map;
+import org.lwjgl.opengl.Display;
+import net.minecraft.client.Minecraft;
+
+public class LabyrinthModNative {
+    public static boolean Loaded = false;
+    private static boolean TriedLoad = false;
+    private static long TimeSinceTried = 0;
+    private static Minecraft mc;
+    private static boolean TryDrawLBMStatus = true;
+
+    public static void Reset() {
+        LabyrinthModNative.Loaded = false;
+        LabyrinthModNative.TimeSinceTried = 0;
+        LabyrinthModNative.TryDrawLBMStatus = true;
+        LabyrinthModNative.TriedLoad = false;
+    }
+
+    public static void Setup(Minecraft mc) {
+        if(LabyrinthModNative.TriedLoad) {
+            return;
+        }
+
+        LabyrinthModNative.mc = mc;
+        
+        try {
+            System.out.println("Looking in "+System.getProperty("java.library.path")+" for libraries");
+            System.loadLibrary("LabyrinthModNative");
+            LabyrinthModNative.init();
+            LabyrinthModNative.Loaded = true;
+        } catch(UnsatisfiedLinkError ex) {
+            System.out.println("Could not load native library! "+ex.toString());
+            LabyrinthModNative.Loaded = false;
+        }
+        LabyrinthModNative.TimeSinceTried = System.currentTimeMillis();
+        LabyrinthModNative.TriedLoad = true;
+    }
+
+    public static native void init();
+
+    public static native void setCoords(double x, double y, double z);
+    public static native void setShouldReset(boolean shouldReset);
+
+    public static void Update() {
+        if(LabyrinthModNative.Loaded) {
+            if(LabyrinthModNative.mc.currentScreen instanceof GuiMainMenu) {
+                LabyrinthModNative.setShouldReset(true);
+                LabyrinthModNative.setCoords(0, 0, 0);
+            } else {
+                LabyrinthModNative.setShouldReset(false);
+                LabyrinthModNative.setCoords(TileEntityRenderer.staticPlayerX, TileEntityRenderer.staticPlayerY, TileEntityRenderer.staticPlayerZ);
+            }
+        }
+    }
+
+    public static void Draw() {
+        try {
+            FontRenderer renderer = mc.fontRenderer;
+
+            String xStr = String.valueOf((int)TileEntityRenderer.staticPlayerX);
+            String yStr = String.valueOf((int)TileEntityRenderer.staticPlayerY);
+            String zStr = String.valueOf((int)TileEntityRenderer.staticPlayerZ);
+
+            if(renderer != null) {
+                renderer.drawStringWithShadow("x: "+xStr, 2, 12, 16777215);
+                renderer.drawStringWithShadow("y: "+yStr, 2, 22, 16777215);
+                renderer.drawStringWithShadow("z: "+zStr, 2, 32, 16777215);
+            }
+            if(LabyrinthModNative.TryDrawLBMStatus) {
+                int lbmPosition = 2;
+                long timeSinceTried = (System.currentTimeMillis() - LabyrinthModNative.TimeSinceTried);
+                if(timeSinceTried >= 5000) {
+                    if(timeSinceTried <= 6000) {
+                        lbmPosition = 2 - ((int)(timeSinceTried - 5000) / 5);
+                    } else {
+                        LabyrinthModNative.TryDrawLBMStatus = false;
+                    }
+                }
+                if(LabyrinthModNative.Loaded) {
+                    renderer.drawStringWithShadow("LabyrinthModNative detected", lbmPosition, 42, 43520);
+                } else {
+                    renderer.drawStringWithShadow("LabyrinthModNative not detected", lbmPosition, 42, 10101010);
+                }
+            }
+        } catch(Exception ex){
+            System.out.println(ex.toString());
+        }
+    }
+
+}
